### ISSUE TITLE
ERROR: Actor crashed! EOFError: end of file reached 	/Users/danolee/.rvm/rubi

### ISSUE THREAD

**Issue by nextofsearch** [alloy/lowdown#14]
Hi,

I am testing the sample code (simple.rb in the examples folder) but getting an error as below. I am using Rails 4.2.5 with ruby 2.3.0p0.

`I, [2016-04-12T17:51:52.625539 #23982]  INFO -- : [APNS Connection#0x7fd8e3140ed8] Connecting...
notification: #<Lowdown::Notification:0x007fd8e21e2f90 @token="XXXXXXXXXXXXXXXXXXXX", @id=1, @payload={:alert=>"Hello HTTP/2! ID=1"}>
W, [2016-04-12T17:51:52.627668 #23982]  WARN -- : [APNS Connection#0x7fd8e3140ed8] Defer performing request, because the connection has not been established yet.
I, [2016-04-12T17:51:53.786826 #23982]  INFO -- : [APNS Connection#0x7fd8e3140ed8] Connection established.
E, [2016-04-12T17:51:54.092417 #23982] ERROR -- : Actor crashed!
EOFError: end of file reached
	/Users/danolee/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/openssl/buffering.rb:178:in `sysread_nonblock'
	/Users/danolee/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/openssl/buffering.rb:178:in `read_nonblock'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:42:in `block in sysread'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:390:in `synchronize'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:40:in `sysread'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:129:in `readpartial'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:223:in `block in runloop'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:221:in `loop'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:221:in `runloop'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
I, [2016-04-12T17:51:54.092732 #23982]  INFO -- : [APNS Connection#0x7fd8e3140ed8] Closing...
[!] Error occurred: end of file reached
Sleep for 5 seconds`

The .pem file and the token are valid ones.

**Comment by nextofsearch** [comment]
I regenerated .pem file, which did not work either. I am using OpenSSL 1.0.2g.

**Comment by nextofsearch** [comment]
`E, [2016-04-13T15:58:54.553659 #3666] ERROR -- : Couldn't cleanly terminate all actors in 10 seconds!
/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `sleep': No live threads left. Deadlock? (fatal)`

**Comment by nextofsearch** [comment]
Also I found this in celluloid issues:
https://github.com/celluloid/celluloid/issues/696

**Comment by alloy** [comment]
If I run the example from this repo, it works fine for me. It appears that you are running the example within a Rails app, so you are probably running into this issue described in the README: https://github.com/alloy/lowdown#gotchas.

**Comment by nextofsearch** [comment]
@alloy I also ran your sample code by 'ruby ./simple.rb' but got the same error.

**Comment by nextofsearch** [comment]
Can you please tell me the versions of your ruby and openssl?

**Comment by nextofsearch** [comment]
I asked my colleague to try out the same code but he also got the same error. He is using ruby 2.1.8 and openssl 1.0.1f

**Comment by alloy** [comment]
Ruby 2.3.0p0 and OpenSSL 1.0.2d_1.

```
$ ruby -v
ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin15]

$ which ruby
/Users/eloy/.rubies/ruby-2.3.0/bin/ruby

$ otool -L /Users/eloy/.rubies/ruby-2.3.0/lib/ruby/2.3.0/x86_64-darwin15/openssl.bundle 
/Users/eloy/.rubies/ruby-2.3.0/lib/ruby/2.3.0/x86_64-darwin15/openssl.bundle:
	/usr/local/opt/openssl/lib/libssl.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/opt/openssl/lib/libcrypto.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	[…]

$ ls -l /usr/local/opt/ | grep openssl
lrwxr-xr-x  1 eloy  admin  26 Nov 20 00:12 openssl -> ../Cellar/openssl/1.0.2d_1
```

**Comment by alloy** [comment]
Btw, I think it’s better to test with the CLI tool. See `$ bundle exec lowdown --help` and let me know how that goes.

**Comment by nextofsearch** [comment]
I got the same error using the CLI tool.

`lowdown -c apns_cert_prod.pem -e production -m "test" $token$
W, [2016-04-13T19:27:52.009082 #13495]  WARN -- : [APNS Connection#0x7fd36920d048] Defer performing request, because the connection has not been established yet.
E, [2016-04-13T19:27:52.768356 #13495] ERROR -- : Actor crashed!
EOFError: end of file reached`

**Comment by nextofsearch** [comment]
I turned on --debug and --verbose options with the CLI tool.

`W, [2016-04-13T19:30:36.077460 #13571]  WARN -- : Dangerously suspending task: type=:call, meta={:dangerous_suspend=>true, :method_name=>:initialize}, status=:callwait
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:41:in `response'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:45:in `value'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/proxy/sync.rb:22:in `method_missing'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/proxy/cell.rb:12:in `_send_'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid.rb:207:in `new_link'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-pool-0.20.5/lib/celluloid/supervision/container/pool.rb:137:in `__spawn_actor__'
I, [2016-04-13T19:30:36.078829 #13571]  INFO -- : [APNS Connection#0x7f8a349291c0] Connecting...
W, [2016-04-13T19:30:36.080872 #13571]  WARN -- : [APNS Connection#0x7f8a349291c0] Defer performing request, because the connection has not been established yet.
I, [2016-04-13T19:30:36.734282 #13571]  INFO -- : [APNS Connection#0x7f8a349291c0] Connection established.
D, [2016-04-13T19:30:36.734352 #13571] DEBUG -- : [APNS Connection#0x7f8a349291c0] Maximum number of streams: 500. Flushing 1 enqueued requests.
D, [2016-04-13T19:30:36.734675 #13571] DEBUG -- : [APNS Connection#0x7f8a349291c0] [00000000-0000-0000-0000-000000000001] Performing request
E, [2016-04-13T19:30:36.930513 #13571] ERROR -- : Actor crashed!
EOFError: end of file reached
	/Users/danolee/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/openssl/buffering.rb:178:in `sysread_nonblock'
	/Users/danolee/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/openssl/buffering.rb:178:in `read_nonblock'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:42:in `block in sysread'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:390:in `synchronize'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:40:in `sysread'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:129:in `readpartial'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:223:in `block in runloop'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:221:in `loop'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:221:in `runloop'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
I, [2016-04-13T19:30:36.930755 #13571]  INFO -- : [APNS Connection#0x7f8a349291c0] Closing...
D, [2016-04-13T19:30:37.063436 #13571] DEBUG -- : Terminating 6 actors...
W, [2016-04-13T19:30:37.064604 #13571]  WARN -- : Dangerously suspending task: type=:finalizer, meta={:dangerous_suspend=>true, :method_name=>:__shutdown__}, status=:receiving
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/actor.rb:230:in `receive'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid.rb:395:in `receive'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/future.rb:90:in `value'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-pool-0.20.5/lib/celluloid/supervision/container/pool.rb:39:in `block in __shutdown__'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-pool-0.20.5/lib/celluloid/supervision/container/pool.rb:39:in `each'
	/Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-pool-0.20.5/lib/celluloid/supervision/container/pool.rb:39:in `__shutdown__'
D, [2016-04-13T19:30:37.064747 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::TerminationRequest:0x007f8a336c7c20>
W, [2016-04-13T19:30:37.064871 #13571]  WARN -- : Terminating task: type=:finalizer, meta={:dangerous_suspend=>true, :method_name=>:__shutdown__}, status=:receiving
	Celluloid::Task::Fibered backtrace unavailable. Please try `Celluloid.task_class = Celluloid::Task::Threaded` if you need backtraces here.
D, [2016-04-13T19:30:37.065143 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::ExitEvent:0x007f8a336c6118>
D, [2016-04-13T19:30:37.065205 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::TerminationRequest:0x007f8a336c7c48>
D, [2016-04-13T19:30:37.065641 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::ExitEvent:0x007f8a336c6730>
D, [2016-04-13T19:30:37.066061 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::ExitEvent:0x007f8a336c6118>
D, [2016-04-13T19:30:37.066142 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::ExitEvent:0x007f8a336c6960>
D, [2016-04-13T19:30:37.066204 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::Call::Sync:0x007f8a336c59e8>
D, [2016-04-13T19:30:37.066486 #13571] DEBUG -- : Discarded message (mailbox is dead): #<Celluloid::Future::Result:0x007f8a3343a368>
/Users/danolee/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/openssl/buffering.rb:178:in `sysread_nonblock': end of file reached (EOFError)
	from /Users/danolee/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/openssl/buffering.rb:178:in `read_nonblock'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:42:in `block in sysread'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:390:in `synchronize'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:40:in `sysread'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-io-0.17.3/lib/celluloid/io/stream.rb:129:in `readpartial'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:223:in `block in runloop'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:221:in `loop'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/lowdown-0.3.1/lib/lowdown/connection.rb:221:in `runloop'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	from /Users/danolee/.rvm/gems/ruby-2.3.0@rails42/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'`

**Comment by alloy** [comment]
It seems like the Apple service closes the connection on you. Can you try the following things?

* Try to use the tool _outside_ of your Rails app context. Things should really work in an isolated context.
* If that still fails, I wonder if there’s something wrong with your cert after all. In case you still have an old-style cert, can you try requesting a new universal cert (that should not invalidate your existing cert) and test with that instead?

**Comment by nextofsearch** [comment]
@alloy I've already tried both the old-style cert and a new style cert. :(

**Comment by nextofsearch** [comment]
@alloy I installed ubuntu on a virtual machine and set up ruby and installed lowdown gem only but got the same result. I got a new push cert again and try it again but got the same result.

**Comment by alloy** [comment]
Wow, that’s so weird. Have you tried other clients, e.g. https://github.com/noodlewerk/NWPusher ?

**Comment by nextofsearch** [comment]
Pusher works fine as long as I uncheck 'Should use sandbox environment'. I used the certificate that showed up in the menu and also tried imported .p12 file. Both work just fine.

I convet .p12 to .pem using this command: 
`openssl pkcs12 -in aps-nudge-demo-new.p12 -out aps-nudge-demo-new.pem  -nodes -clcerts`

Is there any other way to convert .p12 to .pem than this?

**Comment by nextofsearch** [comment]
@alloy I confirmed that my .pem file is valid with multiple services :(

**Comment by patrickslin** [comment]
I had a similar issue. It looks like it is an issue in the http-2 gem dependency. I reverted from version 0.8.1 to 0.8.0 and it fixed the issue.

**Comment by alloy** [comment]
Oooohhhhh, @patrickslin thanks for the feedback!

**Comment by alloy** [comment]
I was indeed able to reproduce it and have pushed a fix, which has been released as version 0.3.2.

**Comment by nextofsearch** [comment]
Thanks @patrickslin and @alloy. It's working now.
